sourced/ml/transformers/tfidf.py

```diff
@@ -0,0 +1,48 @@
+from sourced.ml.algorithms import logTF_log_IDF
+from sourced.ml.transformers.transformer import Transformer
+
+
+class TFIDF(Transformer):
+    """
+    Converts a single UAST into the weighted set (dictionary), where elements are strings
+    and the values are floats. The derived classes must implement uast_to_bag().
+    """
+    DEFAULT_DOCFREQ_THRESHOLD = 5
+
+    def __init__(self, tf, df, docfreq_threshold=None, **kwargs):
+        """
+        :param docfreq_threshold: The minimum number of occurrences of an element to be included \
+                                  into the bag
+        :param weight: TF-IDF will be multiplied by this weight to change importance of specific \
+                      bag extractor
+        """
+        super().__init__(**kwargs)
+
+        self.tf = tf
+        self.df = df
+        self.docfreq_threshold = docfreq_threshold if docfreq_threshold is not None \
+            else self.DEFAULT_DOCFREQ_THRESHOLD
+
+    @property
+    def docfreq_threhold(self):
+        return self._docfreq_threshold
+
+    @docfreq_threhold.setter
+    def docfreq_threshold(self, value):
+        if not isinstance(value, int):
+            raise TypeError("docfreq_threshold must be an integer, got %s" % type(value))
+        if value < 1:
+            raise ValueError("docfreq_threshold must be >= 1, got %d" % value)
+        self._docfreq_threshold = value
+
+    def _get_log_name(self):
+        return type(self).__name__
+
+    def __call__(self, head):
+        ndocs = self.df.count()
+        tfidf = self.df \
+            .map(lambda x: (x[0][1], (x[0][0], x[1]))) \
+            .join(self.tf) \
+            .map(lambda x: (x[1][0][0], x[0], logTF_log_IDF(x[1][1], x[1][0][1], ndocs)))
```

Comment:
At this time I cry blood. Is there any chance you make it accept and RDD of `Row`-s and use the names instead? Or `namedtuple`-s. Otherwise completely unmaintainable.